Fix various recursion issues in Add assumptions

asmeurer opened the pull request:
These only show up when running with the cache off, as otherwise the caching
prevents infinite recursion.

Unfortunately, I don't know how to test this, aside from running the test
suite with the cache off (which we should be doing on Travis or somewhere but
aren't). There are also a lot of other assumedly unrelated test failures
with ./bin/test -C at the moment, although you can see an improvement as
'./bin/test -C core' crashes without this patch and doesn't with it.

Fixes #12907.

asmeurer committed:
Fix various recursion issues in Add assumptions

These only show up when running with the cache off, as otherwise the caching
prevents infinite recursion.

Unfortunately, I don't know how to test this, aside from running the test
suite with the cache off (which we should be doing on Travis or somewhere but
aren't). There are also a lot of other assumedly unrelated test failures
with ./bin/test -C at the moment, although you can see an improvement as
'./bin/test -C core' crashes without this patch and doesn't with it.

Fixes #12907.

asmeurer commented on sympy/core/add.py:
```diff
@@ -576,6 +580,8 @@ def _eval_is_positive(self):
                 s = v + c
+                if s == self:
+                    return None
                 if s.is_positive and a.is_nonnegative:
```
The check here is wrong (we are testing both `s.is_positive` and `a.is_negative`, so there isn't necessarily recursion). 

But I'm having a hard time understanding the `_monotonic_sign` logic. @smichr can you explain what this logic is supposed to do, and what the correct behavior should be if `s` is the same as `self`?

smichr replied:
I would just be parroting back the docstring: the basic idea is to return something indicative of the sign when all symbols are given a signed value/symbol consistent with their assumptions. If s is the same as self it means that the symbol-containing part came back unchanged from `_monotonic_sign`. What sort of expression is giving a recursion error/problem?

asmeurer replied:
You can see in the other assumptions, I had to add checks if `s == self`, or else it recurses trivially (the cache prevents it from being infinite). But here the check also does something with `a`, and I can't figure out what the correct logic is. I believe if you remove the above `s == self` block and try to import with the cache off (`SYMPY_USE_CACHE=no`), you will see the issue.

asmeurer replied:
I really am still quite baffled by what this logic does, but I guess if `s == self` it should skip the whole monotonic_sign stuff, but instead of returning None, continue to the logic below it.

jksuom replied:
It might be expedient to add a test before the `s == self` line. This will be the case when `v == a` is an `_eps` created by `_monotonic_sign`. Its sign can be tested, and if the constant part `c` happens to be non-negative, the sign of `s = v + c` can be determined:
```
    if v.is_positive and c.is_nonnegative:
        return True
```

asmeurer replied:
See the latest change. Is this what you meant?

asmeurer commented:
I would like to get this figured out for the 1.1.1 bugfix release if possible.

asmeurer committed:
Better logic for Add assumptions recursion

When the monotonic_sign stuff would lead to a recursion (only happens when the
cache if off, see #12907, #12908), instead of returning None, skip to the
other logic later in the _eval_is_assumption routine.

jksuom commented:
It looks like the bug in `_eval_is_zero` was causing the infinite recursion. That occurs when some physical units are imported with cache off.
```
In [1]: import sympy

In [2]: from sympy.physics.units import planck_time
...
```
The relevant line is

    planck_time = Quantity("planck_time", time, sqrt(hbar*G/speed_of_light**5), "t_P")

and the error occurs when `sqrt` tries to find the sign of its argument. The assumption system looks into various prerequisite properties of the factors `hbar`, `G` and `speed_of_light**-5` including `algebraic`. It is the last one that leads to infinite recursion on [this line](https://github.com/sympy/sympy/blob/master/sympy/core/power.py#L1065).
```

> /home/jks/sympy/sympy/core/power.py(1065)_eval_is_algebraic()
-> if self.base.is_zero or (self.base - 1).is_zero:
(Pdb) d
> /home/jks/sympy/sympy/core/assumptions.py(247)getit()
-> return _ask(fact, self)
(Pdb) a
self = speed_of_light - 1
(Pdb) p fact
'zero'
(Pdb) d
> /home/jks/sympy/sympy/core/assumptions.py(290)_ask()
-> a = evaluate(obj)
(Pdb) 
> /home/jks/sympy/sympy/core/add.py(542)_eval_is_zero()
-> if self.func(*nz).is_zero:
(Pdb) 
> /home/jks/sympy/sympy/core/assumptions.py(247)getit()
-> return _ask(fact, self)
(Pdb) a
self = speed_of_light - 1
(Pdb) p fact
'zero'
(Pdb) d
> /home/jks/sympy/sympy/core/assumptions.py(290)_ask()
-> a = evaluate(obj)
(Pdb) 
> /home/jks/sympy/sympy/core/add.py(542)_eval_is_zero()
-> if self.func(*nz).is_zero:
(Pdb) 
> /home/jks/sympy/sympy/core/assumptions.py(247)getit()
-> return _ask(fact, self)
(Pdb) a
self = speed_of_light - 1
(Pdb) p fact
'zero'
(Pdb) 
```
This PR will fix the bug but the value of `speed_light_1 - 1` remains undecided.
```
In [1]: import sympy

In [2]: from sympy.physics.units import speed_of_light

In [3]: (speed_of_light - 1).is_zero

In [4]:
```
It seems that the assumptions of class `Quantity` should be improved.

asmeurer commented:
I don't think the old assumptions are capable of handling things like `(x - 1).is_zero`.

asmeurer commented:
In fact, that line could be changed to `if base == 1`, and I suspect nothing would change. There's no way to otherwise assume that something equals 1 in the old assumptions.

asmeurer commented:
I guess is_zero also does some evalf checks, but those can only return a negative result.

jksuom commented:
> In fact, (self.base - 1).is_zero could be changed to self.base == 1

This was exactly what I had in mind, too.

jksuom commented:
One of the assumptions of `Quantity` is `is_real = True`. I feel that this is not right. Reals should be pure numbers.

asmeurer commented:
It probably won't be easy to change that, as I suspect different units rely on being set as nonnegative so that `sqrt(unit**2)` simplifies.
